src/node_http2_core-inl.h

```diff
@@ -193,7 +193,8 @@ inline ssize_t Nghttp2Session::OnStreamReadFD(nghttp2_session* session,
   uv_fs_t read_req;
 
   if (length > 0) {
-    numchars = uv_fs_read(handle->loop_,
+    // TODO(addaleax): Never use synchronous I/O on the main thread.
```

Comment:
re: the todo....

just keep in mind that nghttp2 *requires* that the data be provided synchronously here...

Reply:
Hm – from the nghttp2 header:

> If the application wants to postpone DATA frames (e.g.,
> asynchronous I/O, or reading data blocks for long time), it is
> achieved by returning :enum:`NGHTTP2_ERR_DEFERRED` without reading
> any data in this invocation.  The library removes DATA frame from
> the outgoing queue temporarily.  To move back deferred DATA frame
> to outgoing queue, call `nghttp2_session_resume_data()`.

It even comes with instructions, so that sounds like it’s definitely supported?

Reply:
yes and no. I was being a bit simplistic. `nghttp2_session_resume_data` triggers *this* callback to be invoked again to resume the data flow. Within this method, when the data is available, it must be copied in synchronously. So to achieve what you're looking for it would require a bit more mechanism... e.g. 

1. attempt read,
   * no data? initiate uv_fs_read, return DEFER ... (wait for Step 2) ... keeping in mind that the buffer pointer nghttp2 hands us will be invalid when this function exits, so we'd need to allocate our own buffer for uv_fs_read.
   * got data? copy into nghttp2 provided buffer ...
      * all done? set EOF flag and return.
      * not all done? repeat step 1
2. uv_fs_read callback calls session_resume_data ... flow returns back to step 1

This can be further complicated using the NO_COPY flag, which causes another callback to be called telling our code to send our buffers without the need to memcpy, but there are some complications with that also.

So, yes, it's certainly *possible*, and *likely desirable*, just non-trivial.

Reply:
A bit more background... nghttp2 calls this method while it is attempting to serialize a DATA frame for transmission. Calls to this method are synchronous when ng's send or send_mem function is called. Essentially, nghttp2 goes through it's queue of pending frames and serializes each. When it comes to this method, it will synchronously call it over and over until it receives either a DEFER, an error, or an EOF response. If it receives a DEFER, it takes that stream temporarily out of it's send queue. Calling session_resume_data inserts it back into the queue, so that the next time send or mem_send is called, serialization of the DATA frame can be attempted again.